internal/backend/registry.go

```diff
@@ -3,30 +3,103 @@ package backend
 import (
 	"context"
 	"fmt"
+	"sort"
 )
 
 var (
-	cryptoRegistry  = map[CryptoBackend]CryptoLoader{}
-	storageRegistry = map[StorageBackend]StorageLoader{}
+	CryptoRegistry  = NewRegistry[CryptoBackend, CryptoLoader]()
+	StorageRegistry = NewRegistry[StorageBackend, StorageLoader]()
 
 	// ErrNotFound is returned if the requested backend was not found.
 	ErrNotFound = fmt.Errorf("backend not found")
 )
 
+type Prioritized interface {
+	Priority() int
+}
+
 // CryptoLoader is the interface for creating a new crypto backend.
 type CryptoLoader interface {
 	fmt.Stringer
+	Prioritized
 	New(context.Context) (Crypto, error)
 	Handles(Storage) error
-	Priority() int
 }
 
 // StorageLoader is the interface for creating a new storage backend.
 type StorageLoader interface {
 	fmt.Stringer
+	Prioritized
 	New(context.Context, string) (Storage, error)
 	Init(context.Context, string) (Storage, error)
 	Clone(context.Context, string, string) (Storage, error)
 	Handles(string) error
-	Priority() int
+}
+
+func NewRegistry[K comparable, V Prioritized]() *Registry[K, V] {
+	return &Registry[K, V]{
+		backends: map[K]V{},
+		nameToBackend: map[string]K{},
+		backendToName: map[K]string{},
+	}
+}
+
+type Registry[K comparable, V Prioritized] struct {
+	backends map[K]V
+	nameToBackend map[string]K
+	backendToName map[K]string
+}
+
+func (r *Registry[K, V]) Register(backend K, name string, loader V) {
+	r.backends[backend] = loader
+	r.nameToBackend[name] = backend
+	r.backendToName[backend] = name
+}
+
+func (r *Registry[K, V]) Backends() []string {
+	var names []string
+	for name := range r.nameToBackend {
+		names = append(names, name)
+	}
+	sort.Strings(names)
+	return names
+}
+
+func (r *Registry[K, V]) Prioritized() []V {
+	bes := make([]V, 0, len(r.backends))
+	for _, be := range r.backends {
+		bes = append(bes, be)
+	}
+	sort.Slice(bes, func(i, j int) bool {
+		return bes[i].Priority() < bes[j].Priority()
+	})
+	return bes
+}
+
+func (r *Registry[K, V]) Get(key K) (V, error) {
+	if be, found := r.backends[key]; found {
+		return be, nil
+	}
+	var zero V
+	return zero, ErrNotFound
+}
+
+func (r *Registry[K, V]) Backend(name string) (K, error) {
+	if name == "gpg" {
+		name = "gpgcli"
+	}
```

Comment:
Why do we have a hardcoded "exception" like that here? Wouldn't it be best to change the name on the backend side?

Reply:
yeah, probably. need to clean this up but it's still hardcoded in a bunch of places.